### ISSUE TITLE
Deprecating tlsMode

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5852]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)| @argonemyth reported|
|-|-|
|Trac ID|trac#5852|
|Type|task|
|Created|2012-08-08 12:51:11Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-tlsmode-5852|

The class attribute `tlsMode` in `twisted.mail.smtp.ESMTPClient` is no longer an functional attribute - the usage of secure transport doesn't depend on the value anymore. It's time to remove it from the class.

Attachments:

* [5852.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/86a39d3ae4cb3ba707dba42461708b3df55af47a.patch) (1661 bytes) - added by moijes12 on 2012-08-15 09:42:15Z - Are you looking for something like this ?
* [5852.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/fbffdf6c9dafb05046280696ce5fdd2c9f3c3e2c.patch) (2513 bytes) - added by moijes12 on 2012-09-04 04:53:53Z - 
* [5852.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/afcbf2460546bd3bd6b9d4e622dcb0f6d92cd35c.patch) (2321 bytes) - added by moijes12 on 2012-09-23 14:57:41Z - 
* [5852.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/1388c4367af103ee07b2b835206eb9327ecd5a5d.patch) (3245 bytes) - added by moijes12 on 2012-09-25 20:42:57Z - 
* [5852.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/e0492b0301456922dcf439560e59d4643e8a78f7.patch) (3446 bytes) - added by moijes12 on 2012-09-26 10:41:58Z - 
* [5852.6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/35da3a81d59f4b1160e9e490cc51877285eb5a3a.patch) (3446 bytes) - added by moijes12 on 2012-09-26 10:53:03Z - 
* [5852.7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/a514061c0476dd42fcc070a2cd897a0fb2eb5034.patch) (3446 bytes) - added by lvh on 2013-01-15 13:24:00Z - Same as last patch, but 12.3 -> 13.0
* [5852.8.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/00a3e32af7f0c79c4098d47931251f412cbbd434.patch) (2725 bytes) - added by rwall on 2013-02-03 06:08:50Z - 
* [5852.9.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/da7/da79479a7e4685873b2166f1bcc5e601e7b33466/ad95faf573623b6fd8d8b5317b69bfe7b2f3f1bb.patch) (4433 bytes) - added by rwall on 2013-02-21 13:49:29Z - Now with a __getattr__ method as suggested by tomprince

<details><summary>Searchable metadata</summary>

```
trac-id__5852 5852
type__task task
reporter__argonemyth argonemyth
priority__low low
milestone__ 
branch__branches_deprecate_tlsmode_5852 branches/deprecate-tlsmode-5852
branch_author__moijes12__rwall moijes12, rwall
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__smtp_ESMTPClient_tlsMode smtp ESMTPClient tlsMode
time__1344430271000000 1344430271000000
changetime__1361549793000000 1361549793000000
version__None None
owner__rwall rwall
cc__namn cc__moijes12 cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)|@moijes12 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @moijes12|
|-|-|

Thanks for working on this.

This patch doesn't accomplish the desired result, which is to emit a deprecation when the `tlsMode` attribute is read or written.  The goal is to warn developers with code that manipulates this attribute so they know the attribute no longer means anything.  Introducing a new API that emits a deprecation warning doesn't accomplish the goal, since existing code won't use that API.

A property can be used to achieve the desired effect for when `tlsMode` is read.  Unfortunately, since it's on a classic class, write properties don't work, so a `__setattr__` hook may be necessary.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)|@moijes12 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="namn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|namn commented|
|-|-|

Line 1246 in ```smtp.py``` should be reflowed to 79 columns.

Line 612 in ```test_smtp.py``` looks redundant. I think line 1529  in that same file is also not needed. Line 1542 in the same file should be reflowed.

I would replace ```is``` with ```has been``` in ```5852.removal```.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @moijes12|
|-|-|

Thanks for your continuing work on this issue.

  1. `warnings.catch_warnings` is not how we unit test deprecations.  Please stick to the convention established in other deprecation unit tests in Twisted.
  1. The implementation of `__setattr__` breaks all attribute setting, completely.
  1. The news fragment can just say that the attribute is deprecated, the version information is not needed (the news fragment will appear in the news file under a section for a particular version of Twisted).

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)|@moijes12 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_8"></a>|@moijes12 commented|
|-|-|

twisted.mail.smtp.ESMTPClient being a classic class, I couldn't use descriptors. So I had to use __setattr__ hook. What I did notice was that the test messages were getting suppressed because of previous function definitions and therefore I had to reset the C{.suppress} . Atleast that is what my understanding of it is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @moijes12|
|-|-|

Thanks.  A few issues remain:

  1. The new test asserts the wrong thing.  Creating an ESMTPClient instance should not cause a deprecation warning, _twisted.mail.smtp.ESMTPClient is deprecated since Twisted 12.2_.  `ESMTPClient` is not, in fact deprecated.  Only the `tlsMode` attribute is deprecated.  This means:
    1. The deprecation warning message needs to be changed
    1. The warning needs to be emitted only if the `tlsMode` attribute is set.  Perhaps it is set as a side effect of instantiating `ESMTPClient` now, but if so that is a problem that needs to be addressed.  Using non-deprecated Twisted APIs must not trigger deprecation warnings.
  1. 12.2 has been released.  The next release will probably be 12.3, so that's the version `tlsMode` needs to be deprecated in.
  1. Pass an argument to the `flushWarnings` call.  Otherwise you flush every single warning that has happened, and you avoid testing that the `stacklevel` argument passed to `warnings.warn` was given the correct value.  The argument is a list of functions which could have triggered the warning.  If the `stacklevel` argument has the wrong value, then the wrong function will be marked as triggering the warning and `flushWarnings` will tell you about this by not returning the warning, so your test can fail.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_10"></a>|@moijes12 removed owner|
|-|-|

tlsMode being deprecated, I've removed the its initialisation from the __init__ method of ESMTPClient. The messages have been updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

This needs to happen after [#3989](https://github.com/twisted/twisted/issues/3989).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh set owner to @lvh|
|-|-|

The bug exarkun mentioned is now closed. The patch looks okay to me, apart from the obvious fact that 12.3 is released now, so it should be deprecated in 13.0. I'll fix that and resubmit for review.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_13"></a>|@lvh removed owner|
|-|-|

New patch added, please review.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

# Code Review

 1. Patch didn't apply cleanly to trunk for me - probably since [#3989](https://github.com/twisted/twisted/issues/3989).
    ```
    $ patch -p0 < ../5852.7.patch
    patching file twisted/mail/smtp.py
    Hunk #4 FAILED at 1247.
    1 out of 4 hunks FAILED -- saving rejects to file twisted/mail/smtp.py.rej
    patching file twisted/mail/test/test_smtp.py
    Hunk #1 FAILED at 6.
    Hunk #2 succeeded at 611 (offset 2 lines).
    Hunk #3 FAILED at 1518.
    2 out of 3 hunks FAILED -- saving rejects to file twisted/mail/test/test_smtp.py.rej
    patching file twisted/mail/topfiles/5852.removal
    ```
 2. There's some trailing Whitespace in the patch.
 3. There are still lines > 80 chars
 4. Remove the word "Test " from test method docstring - instead just "ESMTPClient.tlsMode is deprecated"
 5. This ticket title is quite misleading

I'll clean up the patch and resubmit.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in [5852.8.patch](None/5852.8.patch)
 1. Patch should now merge cleanly to trunk
 2. Re-flowed long lines
 3. Moved deprecation test to alongside existing ESMTPClient TLS tests.
 4. Removed unused warning module in test_smtp.py
 5. Removed the "suppress" attribute from the test method - it looked unnecessary.
 6. Added epytext ref to the test docstring.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince set owner to @wallrj|
|-|-|

This only emits a deprecation warning when writing to `tlsMode`, not when reading. Both should be handled. Otherwise, this looks good.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj removed owner|
|-|-|

Replying to [tom.prince](#note_16):
> This only emits a deprecation warning when writing to `tlsMode`, not when reading. Both should be handled. Otherwise, this looks good.

Ok. 
 * I renamed the actual tlsMode value to _tlsMode
 * I added a __getattr__ method and modified __setattr__ to access the private var
 * ...and some new tests.

I didn't try and do anything about accessing or setting it as a class variable

See [5852.9.patch](None/5852.9.patch)

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Putting this in a branch and see what the buildfarm thinks of it.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra commented|
|-|-|

(In [37187]) Branching to 'deprecate-tlsmode-5852'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra commented|
|-|-|

(In [37188]) Apply 5852.9.patch, refs [#5852](https://github.com/twisted/twisted/issues/5852)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_21"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-tlsmode-5852).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince set owner to @wallrj|
|-|-|

One minor point:
  * The `AttributeError` should use `self.__class__.__name__` instead of hardcoding it.
Otherwise, this looks good. Thanks for your work on this.
Please commit after fixing that.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_23"></a>|@wallrj set status to `closed`|
|-|-|

(In [37202]) Merge deprecate-tlsmode-5852: tlsMode attribute of twisted.mail.smtp.ESMTPClient is deprecated.

Author: moijes12, rwall
Reviewer: exarkun, namn, lvh, rwall, tom.prince, thijs
Fixes: [#5852](https://github.com/twisted/twisted/issues/5852)

The tlsMode attribute of twisted.mail.smtp.ESMTPClient is deprecated.
